### DISCUSSION TITLE
Implementation for waveguide terminations

### DISCUSSION THREAD

**Discussion by gadiLhv** [thliebig/openEMS-Project#107] [General]
Hello there

I have noticed that in the rectangular waveguide example, the termination is obtained by placing a PML layer at the "receiving" end.
Digging down a bit I did not find an implementation for waveguide terminations to build the receiving ports.
 
Is there an implementation that I missed for this?

**Comment by thliebig** [comment]
I'm not sure that I understand what you mean? What is a receiving end of a waveguide?
A waveguide should always terminate in something like a PML to make sure reflections into the waveguide are absorbed.
If a port is active or not is only determined if the excitation is on or off (not existent) .
Additionally a rect waveguide is a waveguide...
I'm really not sure what you want to know...

**Reply by gadiLhv** [reply]
The excitation is a bit different, as there is no "resistor" to absorb the reflected wave, especially in the single conductor case (e g. rectangular WG). I'm still figuring this part out.

In the receive ports, an absorbing B.C or PML are a natural solutions, I don't want to start with multimodal absorption, yet. 

My question would be if openEMS supports "local" B.C. definitions, in cases that the waveguide termination is not on the edge of the simulation domain.

**Reply by thliebig** [reply]
No openEMS does not currently support local PML, thus any waveguide port (excited or not) has to terminate at the boundary.
I'm still confused about the "receive" parts.... Any port needs to absorb incoming power somehow. It has nothing to do with receive or transmit, active or passive...

**Reply by gadiLhv** [reply]
Sorry for taking so long to respond.

I think this example will clarify what I meant. 

Currently, the 'Rect_Waveguide.m' example yields the following results:
https://github.com/thliebig/openEMS-Project/assets/3908295/971cc674-b43d-4329-90a0-46f4acedd364

I would expect that if I defined a non-exciting ports, as I did here
```matlab
start=[mesh.x(1)   mesh.y(1)   mesh.z(end-5) ];
stop =[mesh.x(end) mesh.y(end) mesh.z(end-5) ];
CSX = AddExcitation(CSX,'receive',0,[0 0 0]);
weight{1} = func_Ex;
weight{2} = func_Ey;
weight{3} = 0;
CSX = SetExcitationWeight(CSX,'receive',weight);
CSX = AddBox(CSX,'receive',0 ,start,stop);
```
That the result would be the same, even without a PML nearby. Expectadly, these are the results obtained:
https://github.com/thliebig/openEMS-Project/assets/3908295/9aec55cf-0144-41cf-b7e4-8611e139b345

My question is, is there an implementation currently in OpenEMS to absorb the outgoing wave.

**Reply by thliebig** [reply]
What you have defined there is not a port, it is just an excitation which seems to excite with 0 V/m, thus doing nothing...

To absorb energy you do not need an excitation but e.g. a PML or an absorbing sheet. But for waveguides and their complex field-distribution, a PML (boundary condition)  is the only proper way to go.
Simple ports like "lumped ports" use a resistive, sheet matched to the device, that can do it, but that does not work to well for waveguides.

**Reply by gadiLhv** [reply]
Yes, I was aware that the example won't obtain absorption, as stated before.

Now I need to do some trial and error with the absorption of the two possible BCs now to decide how to implement it.

**Reply by gadiLhv** [reply]
So while I'm trying to implement the "localized" boundary conditions, I found this phoenomenon:

If I try the rectangular waveguide example with the original boundary conditions, it works fine and well.

However, if I try and back the excitation port with a PML also by setting
```matlab
BC = [0 0 0 0 3 3];
```
and shifting the excitation slightly away
```matlab
start=[mesh.x(1)   mesh.y(1)   mesh.z(2) ];
stop =[mesh.x(end) mesh.y(end) mesh.z(2) ];
CSX = AddExcitation(CSX,'excite',0,[1 1 0]);
```
The simulation doesn't converge. Looking at the animation showed something that looked like an evanescent mode that isn't absorbed.

https://github.com/thliebig/openEMS-Project/assets/3908295/f95f1863-cc96-4283-b9a3-62b0b1228e36

Do you have any insights regarding why this is happening and how to avoid it?

Also, I am trying to implement the "super-absorbing" boundary conditions suggested in [1] and [2] below, for the localized absorbers.
If you have any previous experience with them, I would love to hear about it.

[1] Fang, J., and Kenneth K. Mei. "A super-absorbing boundary algorithm for solving electromagnetic problems by time-domain finite-difference method." 1988 IEEE AP-S. International Symposium, Antennas and Propagation. IEEE, 1988.
‏
[2] Mei, Kenneth K., and Jiayuan Fang. "Superabsorption-a method to improve absorbing boundary conditions (electromagnetic waves)." IEEE Transactions on Antennas and Propagation 40.9 (1992): 1001-1010.‏

**Reply by gadiLhv** [reply]
I've managed to write most of the operator and engine extensions for this, but before finalizing, I managed to answer one of my questions.

I've used some old home-work of mine to test the benefit of the "superabsorbing" (I didn't give it that name, nor would I, for real) boundary conditions (SABC).

**The first example** set is for a mode propagatin in a homogenous waveguide. Note that this is a 2D formulation, so only Ex,Ey and Hz are solved here.

Here is the peak level of Ey field (in logarithmic scale) for the two solutions
![Mur_Vs_SABC](https://github.com/thliebig/openEMS-Project/assets/3908295/c8b002d5-beeb-4f2d-83ab-fb85d67709c5)

Without SABC (1st order Mur):
https://github.com/thliebig/openEMS-Project/assets/3908295/bdf1bdbe-f13a-4114-95bc-3fb60ff39418

With SABC:
https://github.com/thliebig/openEMS-Project/assets/3908295/40106a16-f119-498a-9895-0098756ad020

**The 2nd case** is a the same waveguide, but with a metalic plate. Here the benefit is less apparent as the decay is much slower (I'm guessing due to reflections), but still visible.

![Mur_Vs_SABC_W_Plate](https://github.com/thliebig/openEMS-Project/assets/3908295/e785ceb2-54b7-4406-9954-9bbf5f89ab6d)

1st order Mur:
https://github.com/thliebig/openEMS-Project/assets/3908295/fd29bcd5-b148-455d-9193-5b3bf24c9ec7

SABC:
https://github.com/thliebig/openEMS-Project/assets/3908295/66f73061-b7cd-4a19-b2ec-81376373c8da

This should be implemented soon enough.
